### ISSUE TITLE
`optax` and `tensorflow`'s Adam optimizer's setting.

### ISSUE THREAD

**Issue by vwxyzjn** [deepmind/optax#571]
Currently, `optax.scale_by_adam` should be equivalent to `torch.optim.Adam`. However, Tensorflow has a different implementation. 

In short, if we change https://github.com/deepmind/optax/blob/cebdeff4a1922113a96c520e7a81b5bf79825b77/optax/_src/transform.py#L345-L348 to the following, then the adam optimizer would be the same as tensorflow's imlementation.

```python
updates = jax.tree_util.tree_map(
    lambda m, v: (jnp.sqrt(1- b2**count_inc) / (1-b1**count_inc)) *  m / (jnp.sqrt(v + eps_root) + eps), mu, nu)
```

### More context

Basically, PyTorch and optax's adam follow Algorithm 1 of the Kingma and Ba’s Adam paper ([arxiv/1412.6980](https://arxiv.org/pdf/1412.6980.pdf)), but TensorFlow uses the formulation just before Section 2.1 of the paper and its epsilon referred to here is epsilon hat in the paper.

This was a relevant issue in my recent reproduction of openai's work in https://github.com/openai/lm-human-preferences. Long story short, below is an end-to-end experiment with torch's adam `adam_pt` and tensorlfow-style adam `adam_tf`. While the final performance (`objective/scores`) look the same, the learning curves are different in a non-trivial way. E.g., the torch adam version had a much higher `clipfrac` initially, causing a more initial significant update. 

<img width="1235" alt="image" src="https://github.com/deepmind/optax/assets/5555347/87222fd1-ff04-482d-a226-4512d39da12b">

The "initial aggressive update" issue gets aggravated in larger models (e.g., gpt2-large). You can see that `objective/kl` had a spike with `adam_tf`, so this could be a reproducibility issue.

<img width="1236" alt="image" src="https://github.com/deepmind/optax/assets/5555347/d14b3f72-67f7-4e25-bbbc-199521e65c64">

### Desired solution

include a 

```python
def scale_by_adam_tfstyle(
    b1: float = 0.9,
    b2: float = 0.999,
    eps: float = 1e-8,
    eps_root: float = 0.0,
    mu_dtype: Optional[chex.ArrayDType] = None,
) -> base.GradientTransformation:
  mu_dtype = utils.canonicalize_dtype(mu_dtype)

  def init_fn(params):
    mu = jax.tree_util.tree_map(  # First moment
        lambda t: jnp.zeros_like(t, dtype=mu_dtype), params)
    nu = jax.tree_util.tree_map(jnp.zeros_like, params)  # Second moment
    return ScaleByAdamState(count=jnp.zeros([], jnp.int32), mu=mu, nu=nu)

  def update_fn(updates, state, params=None):
    del params
    mu = update_moment(updates, state.mu, b1, 1)
    nu = update_moment_per_elem_norm(updates, state.nu, b2, 2)
    count_inc = numerics.safe_int32_increment(state.count)
    updates = jax.tree_util.tree_map(
        lambda m, v: (jnp.sqrt(1- b2**count_inc) / (1-b1**count_inc)) *  m / (jnp.sqrt(v + eps_root) + eps), mu, nu) # 
    mu = utils.cast_tree(mu, mu_dtype)
    return updates, ScaleByAdamState(count=count_inc, mu=mu, nu=nu)

  return base.GradientTransformation(init_fn, update_fn)
```

obviously this is bad naming, but I figure you might do some refactoring :)

**Comment by mtthss** [comment]
Thanks for raising this!

Rather than including an alternative adam version in optax, I would instead suggest we add a warning in the doc-string of the adam and scale_by_adam gradient transformations. Once they are aware, it's easy for people who need to reproduce old results to fork and modify the transformation.

E.g. we could add:
```
WARNING: PyTorch and optax's adam follow Algorithm 1 of the Kingma and Ba’s Adam paper, if reproducing old results note that TensorFlow used instead the formulation just before Section 2.1 of the paper.
```

Can you put together a short PR with this fix?

**Comment by vwxyzjn** [comment]
@mtthss with pleasure. Thanks for the suggestion and see https://github.com/deepmind/optax/pull/572. 

P.S. big fan of your work and it made my day seeing your response :D

**Comment by liutianlin0121** [comment]
Hi @vwxyzjn @mtthss ,

Thanks Costa @vwxyzjn for the observation regarding Optax vs. Tensorflow Adam and for bringing this issue to my attention.

>While the final performance (objective/scores) look the same, the learning curves are different in a non-trivial way. E.g., the torch adam version had a much higher clipfrac initially, causing a more initial significant update.

> PyTorch and optax's adam follow Algorithm 1 of the Kingma and Ba’s Adam paper ([arxiv/1412.6980](https://arxiv.org/pdf/1412.6980.pdf)), but TensorFlow uses the formulation just before Section 2.1 of the paper and its epsilon referred to here is epsilon hat in the paper.

Here are derivations that verify Costa's observations mentioned above. Following the notation of Kingma and Ba’s Adam paper, let's compare the update equations implemented by Optax and Tensorflow. We have the gradient update rules for optax/pytorch adam (Algorithm 1 of Kingma and Ba’s paper) and tensorflow adam (the formulation just before Section 2.1 of Kingma and Ba’s paper) as below.

```math
\begin{equation}
\begin{aligned}\text{optax/pytorch adam :}\quad 
\theta_t & =\theta_{t-1}-\alpha \cdot \hat{m}_t /\left(\sqrt{\hat{v}_t}+\varepsilon\right) \\
& =\theta_{t-1}- \alpha \underbrace{\left[m_t /\left(1-\beta_1^t\right)\right]}_{=\hat{m}_t} /\left[\sqrt{\underbrace{v_t /\left(1-\beta_2^t\right)}_{=\hat{v}_t} }+\varepsilon\right]\\
& =\theta_{t-1}- \alpha\left[m_t /\left(1-\beta_1^t\right)\right]\frac{\sqrt{1-\beta_2^t}}{\sqrt{v_t}+\color{blue}{\varepsilon \sqrt{1-\beta_2^t}}}
\end{aligned}
\end{equation}
```

```math
\begin{equation} \begin{aligned}\text{tensorflow adam:}\quad \theta_t & =\theta_{t-1}-\alpha_t m_t /\left(\sqrt{v_t}+\hat{\varepsilon}\right) \\& =\theta_{t-1}-\underbrace{\left[\alpha \sqrt{1-\beta_2^t} /\left(1-\beta_1^t\right)\right]}_{=\alpha_t} m_t /\left(\sqrt{v_t}+\hat{\varepsilon}\right) \\& =\theta_{t-1}- \alpha\left[m_t /\left(1-\beta_1^t\right)\right] \frac{\sqrt{1-\beta_2^t}}{\sqrt{v_t}+\color{blue}{\hat{\varepsilon}}} \end{aligned}\end{equation}
```

The equations above highlight that the distinction between optax and tensorflow implementation is their normalization terms, $\color{blue}{\varepsilon \sqrt{1-\beta_2^t}}$  and $\color{blue}{\hat{\varepsilon}}$. The two versions are equivalent if we set  $\hat{\varepsilon}  =\varepsilon \sqrt{1-\beta_2^t}$ .  However, in the optax and tensorflow APIs, we can only set $\varepsilon$ (optax and pytorch) and $\hat{\varepsilon}$ (tensorflow) via the `eps` argument, causing differences in their update equations. 

So, what if we set  $\varepsilon$ and $\hat{\varepsilon}$  to the same value, say, 1e-5? Then for tensorflow adam, the normalization term $\hat{\varepsilon} = \text{1e-5}$  is just a constant. But for optax/pytorch adam, the normalization term ${\varepsilon \sqrt{1-\beta_2^t}}$ changes over time. Importantly, the normalization term  ${\varepsilon \sqrt{1-\beta_2^t}}$ is initially much smaller than 1e-5 when the timestep $t$ is small, and gradually approaches to 1e-5 as timesteps increase. The plot below compares these two normalization terms over timesteps:

<img src="https://github.com/deepmind/optax/assets/10226549/13967b2a-a4bd-4083-9443-223a2ff7da8d" alt="image" width="60%" height="auto" class="center">


The above figure shows that, if we set the same `eps` in optax adam and tensorflow adam, then optax/pytorch adam uses a much smaller normalization term than tensorflow-adam in the early phase of training. In other words, optax/pytorch adam goes for more aggressive gradient updates early in the training. This aligns nicely with Costa's empirical results shared earlier!

Tianlin
